webrender/res/brush.glsl

```diff
@@ -44,36 +56,79 @@ void main(void) {
     BrushInstance brush = load_brush();
 
-    // Fetch the dynamic picture that we are drawing on.
-    PictureTask pic_task = fetch_picture_task(brush.picture_address);
-
     // Load the geometry for this brush. For now, this is simply the
     // local rect of the primitive. In the future, this will support
     // loading segment rects, and other rect formats (glyphs).
     PrimitiveGeometry geom = fetch_primitive_geometry(brush.prim_address);
 
-    // Write the (p0,p1) form of the primitive rect and the local position
-    // of this vertex. Specific brush shaders can use this information to
-    // interpolate texture coordinates etc.
-    vLocalRect = vec4(geom.local_rect.p0, geom.local_rect.p0 + geom.local_rect.size);
-
-    // Right now - pictures only support local positions. In the future, this
-    // will be expanded to support transform picture types (the common kind).
-    vec2 pos = pic_task.target_rect.p0 + aPosition.xy * pic_task.target_rect.size;
-    vLocalPos = aPosition.xy * pic_task.target_rect.size / uDevicePixelRatio;
+    vec2 device_pos, local_pos;
+    RectWithSize local_rect = geom.local_rect;
+
+    if ((brush.flags & BRUSH_FLAG_USES_PICTURE) != 0) {
+        // Fetch the dynamic picture that we are drawing on.
+        PictureTask pic_task = fetch_picture_task(brush.picture_address);
+
+        // Right now - pictures only support local positions. In the future, this
+        // will be expanded to support transform picture types (the common kind).
+        device_pos = pic_task.target_rect.p0 + aPosition.xy * pic_task.target_rect.size;
+        local_pos = aPosition.xy * pic_task.target_rect.size / uDevicePixelRatio;
+
+        // Write the final position transformed by the orthographic device-pixel projection.
+        gl_Position = uTransform * vec4(device_pos, 0.0, 1.0);
+    } else {
+        AlphaBatchTask alpha_task = fetch_alpha_batch_task(brush.picture_address);
+        Layer layer = fetch_layer(brush.layer_address);
+        ClipArea clip_area = fetch_clip_area(brush.clip_address);
+
+        // Write the normal vertex information out.
+        // TODO(gw): Support transform types in brushes. For now,
+        //           the old cache image shader didn't support
+        //           them yet anyway, so we're not losing any
+        //           existing functionality.
+        VertexInfo vi = write_vertex(
+            geom.local_rect,
+            geom.local_clip_rect,
+            float(brush.z),
+            layer,
+            alpha_task,
+            geom.local_rect
+        );
+
+        local_pos = vi.local_pos;
+
+        // For brush instances in the alpha pass, always write
+        // out clip information.
+        // TODO(gw): It's possible that we might want alpha
+        //           shaders that don't clip in the future,
+        //           but it's reasonable to assume that one
+        //           implies the other, for now.
+#ifdef WR_FEATURE_ALPHA
+        write_clip(
```

Comment:
could technically use a `BRUSH_FLAG_XXX` for that

Reply:
I think it makes sense to have separate shaders for this (since it affects the FS too), but we could certainly revisit later on after profiling.